### Importing Libraries

In [189]:
# Importing the required libraries

import pandas as pd
from numpy import *
import numpy as np
from sklearn import preprocessing
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn import metrics
from sklearn import svm
from sklearn import neighbors

from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report

### Importing and pre-processing the data


In [190]:
# Load the training dataset

data =pd.read_csv('train dataset.csv')
data.head() # To display first 5 records in training datset

Gender  Age  openness  neuroticism  conscientiousness  agreeableness  \
0    Male   17         7            4                  7              3   
1    Male   19         4            5                  4              6   
2  Female   18         7            6                  4              5   
3  Female   22         5            6                  7              4   
4  Female   19         7            4                  6              5   

   extraversion Personality (Class label)  
0             2               extraverted  
1             6                   serious  
2             5                dependable  
3             3               extraverted  
4             4                    lively

In [191]:
# Converting categorical data into numerical.
# Male: 1 and Female: 0
data=data.replace(to_replace=['Male', 'Female'], value=[1,0])
data.head()

Gender  Age  openness  neuroticism  conscientiousness  agreeableness  \
0      1   17         7            4                  7              3   
1      1   19         4            5                  4              6   
2      0   18         7            6                  4              5   
3      0   22         5            6                  7              4   
4      0   19         7            4                  6              5   

   extraversion Personality (Class label)  
0             2               extraverted  
1             6                   serious  
2             5                dependable  
3             3               extraverted  
4             4                    lively

In [192]:
# Divide training data based on independent and dependent attributes

array = data.values
df=pd.DataFrame(array)
maindf =df[[0,1,2,3,4,5,6]] 
train_x=maindf.values
print (train_x) # independent variables 


[[1 17 7 ... 7 3 2]
 [1 19 4 ... 4 6 6]
 [0 18 7 ... 4 5 5]
 ...
 [1 22 5 ... 3 6 1]
 [1 19 5 ... 5 7 5]
 [0 18 5 ... 7 6 5]]


In [193]:
temp=df[7] # target variable
train_y =temp.values

for i in range(len(train_y)):
	train_y[i] =str(train_y[i])
print(train_y) # dependent/target variables

['extraverted' 'serious' 'dependable' 'extraverted' 'lively' 'lively'
 'extraverted' 'extraverted' 'dependable' 'lively' 'serious' 'extraverted'
 'lively' 'dependable' 'responsible' 'extraverted' 'serious' 'dependable'
 'responsible' 'extraverted' 'responsible' 'extraverted' 'responsible'
 'serious' 'serious' 'dependable' 'extraverted' 'extraverted' 'lively'
 'lively' 'extraverted' 'extraverted' 'dependable' 'lively' 'serious'
 'lively' 'responsible' 'responsible' 'dependable' 'serious' 'dependable'
 'extraverted' 'serious' 'extraverted' 'dependable' 'dependable'
 'responsible' 'serious' 'lively' 'lively' 'extraverted' 'extraverted'
 'serious' 'extraverted' 'lively' 'lively' 'extraverted' 'dependable'
 'lively' 'lively' 'responsible' 'extraverted' 'responsible' 'dependable'
 'serious' 'serious' 'responsible' 'dependable' 'responsible' 'dependable'
 'extraverted' 'extraverted' 'responsible' 'dependable' 'responsible'
 'serious' 'extraverted' 'dependable' 'lively' 'lively' 'extraverted'


### Training the model

In [194]:
# Using Logistic Regression

model_logistic = linear_model.LogisticRegression(multi_class='multinomial', solver='newton-cg',max_iter =1000)
model_logistic.fit(train_x, train_y) 

LogisticRegression(max_iter=1000, multi_class='multinomial', solver='newton-cg')

In [195]:
# Using Support Vector Machine

model_svm=svm.SVC(kernel="poly",C=5)
# Perform the training
model_svm.fit(train_x,train_y)

SVC(C=5, kernel='poly')

### Predicting Target variable using Test data

In [196]:
# Load test dataset

testdata =pd.read_csv('test dataset.csv')
test = testdata.values

In [197]:
# converting categorical data into numerical.
# Male: 1 and Female: 0

for i in range(len(test)):
	if test[i][0]=="Male":
		test[i][0]=1
	else:
		test[i][0]=0

In [198]:
# Store independent variable values in test_x array

df1=pd.DataFrame(test)
testdf =df1[[0,1,2,3,4,5,6]]
test_x=testdf.values
print(test_x)

[[0 20 7 ... 9 5 5]
 [1 17 5 ... 5 2 4]
 [0 25 5 ... 7 2 4]
 ...
 [1 18 7 ... 6 2 7]
 [1 23 6 ... 5 4 3]
 [0 18 5 ... 3 5 6]]


In [199]:
# store target variable values in test_y array

temp=df1[7]
test_y =temp.values

In [200]:
# Predicting target attribute  value by using logistic regression model

y_pred_logistic = model_logistic.predict(test_x)

# Store predicted target variable values in ouput_logistic.csv file

for i in range(len(y_pred_logistic)) :
	y_pred_logistic[i]=str((y_pred_logistic[i]))
DF = pd.DataFrame(y_pred_logistic,columns=['Predicted Personality'])
DF.index=DF.index+1
DF.index.names = ['Person No']
DF.to_csv("output_logistic.csv")

In [201]:
# Predicting target attribute  value by using SVM model

y_pred_svm= model_svm.predict(test_x)

# Store predicted target variable values in ouput_logistic.csv file

for i in range(len(y_pred_svm)) :
	y_pred_svm[i]=str((y_pred_svm[i]))
DF = pd.DataFrame(y_pred_logistic,columns=['Predicted Personality'])
DF.index=DF.index+1
DF.index.names = ['Person No']
DF.to_csv("output_svm.csv")

### Evaluating 

#### Logistic Regression

In [202]:
# Calculating accuracy of Logistic model
accuracy_logistic = model_logistic.score(test_x,test_y)
print(accuracy_logistic)

0.7968253968253968


#### SVM

In [205]:
# Calculating accuracy of SVM model
accuracy_svm = model_svm.score(test_x,test_y)
print(accuracy_svm)

0.5873015873015873


### Conclusion

In [208]:
# Classification report for Logistic Regression:
print(classification_report(test_y,y_pred_logistic))

              precision    recall  f1-score   support

  dependable       0.33      0.24      0.28        21
 extraverted       0.80      0.86      0.83        77
      lively       0.88      0.62      0.73        24
 responsible       0.93      0.65      0.76        40
     serious       0.80      0.91      0.85       153

    accuracy                           0.80       315
   macro avg       0.75      0.66      0.69       315
weighted avg       0.79      0.80      0.79       315



In [209]:
# Classification report for SVM:
print(classification_report(test_y,y_pred_svm))

              precision    recall  f1-score   support

  dependable       0.14      0.05      0.07        21
 extraverted       0.63      0.52      0.57        77
      lively       0.18      0.08      0.11        24
 responsible       0.56      0.12      0.20        40
     serious       0.61      0.90      0.72       153

    accuracy                           0.59       315
   macro avg       0.42      0.33      0.34       315
weighted avg       0.54      0.59      0.53       315



Accuracy of model when trained with <b>Logistic Regression Algortihm: 79.68 %</b><br>
Accuracy of model when trained with <b>SVM: 58.73 %</b>
<hr>
Therfore Logistic regression works well for the given problem.